### PR TITLE
Use implicit (not explicit) rules for promotability by default in `const fn`

### PR DESCRIPTION
**Pull request by ecstatic-morse**
For crater run. See https://github.com/rust-lang/const-eval/pull/54#discussion_r469995552.

cc #75586

### CODE REVIEW

**Comment by ecstatic-morse** [comment]
@craterbot check

cc @RalfJung. Unfortunately the queue is long.

File: src/librustc_mir/transform/promote_consts.rs [hunk 1]
```diff
@@ -748,7 +748,14 @@ impl<'tcx> Validator<'_, 'tcx> {
...
 
-        if !self.explicit && self.const_kind.is_none() {
+        // `const` and `static` use the explicit rules for promotion, meaning calls to `const fn`
+        // can be promoted.
+        let context_uses_explicit_promotion_rules = matches!(
+            self.const_kind,
+            Some(hir::ConstContext::Static(_) | hir::ConstContext::Const)
+        );
+
+        if !self.explicit && !context_uses_explicit_promotion_rules {
```
Comment (RalfJung) [inline src/librustc_mir/transform/promote_consts.rs:758]: `self.explicit` is already encoding whether we use explicit promotion, and that's what we check everywhere else in this file. Why does this one check need special treatment?

**Comment by RalfJung** [comment]
Let's also cc @oli-obk.

---

I must still be missing something. This line sets implicit promotion rules for lifetime extension:

https://github.com/rust-lang/rust/blob/860d599d5b11db275a829a3f898560405fa84bbe/src/librustc_mir/transform/promote_consts.rs#L801

Where is the code that overwrites this to use explicit promotion rules for lifetime extension in `const`/`static` items and `const fn` bodies?

File: src/librustc_mir/transform/promote_consts.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_mir/transform/promote_consts.rs:758]: Ah I see, it's the `self.const_kind.is_none()` here that introduced the exception. IMO that should just be `if !self.explicit`... but that doesn't seem how we currently organize things.

Comment (RalfJung) [reply src/librustc_mir/transform/promote_consts.rs:758]: However, this means that all `const_kind.is_none()` are wrong. Promotion should behave the same for `const_kind == Some(ConstFn)` and `const_kind == None`.

**Comment by RalfJung** [comment]
Give the length of the queue, I wonder if it's not worth to test all the cases there `const fn` are treated differently from `fn` at once. Those are the most critical points, I think.

Btw the PR title needs adjusting -- this is not about explicit vs implicit promotion (as witnessed by the code), it's about promotion depending on `const_kind`.

File: src/librustc_mir/transform/promote_consts.rs (hunk 1 above)
Comment (ecstatic-morse) [reply src/librustc_mir/transform/promote_consts.rs:758]: Most of these checks are for things that are not allowed in a const context (without feature gates). Presumably this logic is left over from when promotion and const-checking happened in the same module.

Comment (RalfJung) [reply src/librustc_mir/transform/promote_consts.rs:758]: Ah, that's a good point.

Still, it wouldn't hurt to adjust them all to call something like `is_const_only` or so (meaning "this code will only ever be executed during CTFE), just to make sure we haven't missed any of them.

**Comment by ecstatic-morse** [comment]
There appears to be a single, legit root regression in the `x86_64` crate when the `const_fn` feature is enabled.

https://github.com/rust-osdev/x86_64/blob/fe8d09c7edeaa399d6390da31de8aa7c4819731a/src/structures/paging/page_table.rs#L190-L194

It seems that we could change these semantics without too severe of an impact on the ecosystem. @RalfJung what's the next step here?

**Comment by RalfJung** [comment]
> There appears to be a single, legit root regression in the x86_64 crate when the const_fn feature is enabled.

That's great news!
I do not even see the use of promotion there? `ENTRY_COUNT` is a `const`.

> It seems that we could change these semantics without too severe of an impact on the ecosystem. @RalfJung what's the next step here?

We nominate this for lang-team discussion and try to get the FCP process started.

**Comment by RalfJung** [comment]
@rust-lang/lang the summary for the FCP is that this does most of what it takes to fix https://github.com/rust-lang/rust/issues/75586. For further information, see that issue; for the big picture, see https://github.com/rust-lang/const-eval/issues/53.

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

As we have discussed in numerous meetings, this is a move to limit const promotion, which overall was much too aggressive. As @RalfJung [wrote](https://github.com/rust-lang/rust/pull/75502#issuecomment-684832815):

> the summary for the FCP is that this does most of what it takes to fix https://github.com/rust-lang/rust/issues/75586. I view this as an instance of "we accidentally stabilized too much". For further information, see that issue; for the big picture, see https://github.com/rust-lang/const-eval/issues/53.

**Comment by RalfJung** [comment]
@ecstatic-morse this just needs a rebase, then we can land it. :)

**Comment by RalfJung** [comment]
@ecstatic-morse you can find a rebased version of this patch at f7e1b1bd63fa21a465d73e000c74564eb68d1e0e. It just needs pushing to this branch then I can r+. :)

**Comment by RalfJung** [comment]
Sorry, I meant 65ee8a787d33dd5a0f0695d81508acdd9e4c8f3c.

**Comment by RalfJung** [comment]
Let's reflect the reviewer.
@bors r- r=RalfJung

**Comment by lcnr** [comment]
This is a breaking change so it probably makes sense to quickly mention this in the release notes

**Pull request merged**